### ISSUE TITLE
[PROPOSAL] Rework and Make Programmatic Names and Namespaces

### ISSUE THREAD

**Issue by denimalpaca** [OpenLineage/OpenLineage#1681]
**Purpose:**
The Naming.md file should be reworked as a more programmatic solution with clear, specific definitions.

Names and Namespaces are currently slightly nebulous concepts. They can best be described, perhaps, by two components that, together, form a unique URI to a specific dataset for datasets, and a unique name for a Job. This is a serviceable definition for databases, datalakes, and distributed file stores, where a well-defined path exists by nature of the structure. But this may not be the case with database-like systems, or non-database systems, like Salesforce or Google Sheets, respectively. These instances need another way to specify a unique name, one that may not resolve to a URI.

Further, the[Naming.md](https://github.com/OpenLineage/OpenLineage/blob/main/spec/Naming.md) has some shortcomings besides support for only databases, datalakes, and distributed file stores:

- Little information, beyond Airflow and Spark, on how Jobs are named within integrations
- Limited set of dataset names, no specific process for adding more
- No definitions of any major terms
- Not programmatic; reference only

There also seems to be inconsistency with naming between certain integrations (at the time of this writing, the SnowflakeExtractor and Great Expectations emit different namespaces and names for Snowflake datasets. This proposal would help rectify issues that cause these discrepancies and ensure that when changes to naming happen, they are consistent across all integrations. The easiest way to do this may be to make things more programmatic, or at least have the ability to test naming changes against integrations.

**Proposed implementation**
The Naming.md file should be transitioned to a set of JSON files that specify the naming conditions for a particular integration or source, which can then be used to build a reference file.

The file hierarchy should look like:
```
spec/
  CONTRIBUTING.md
  naming/
    jobs/
      airflow/
        airflow.json
      ...
    datasets/
      snowflake/
        snowflake.json
      ...
```

Where CONTRIBUTING.md (or something like it) would take the place of the current Naming.md's for purposes of explaining naming philosophy and how to add or update files in the new file structure, and Naming.md would transition to be a reference file built from the JSON files under `naming/`. CONTRIBUTING.md would also provide clear definitions for all major terms, including: name, namespace, datasource hierarchy, naming hierarchy, scheme, authority, 

Each file under `naming/` would specify the particular convention for that integration or dataset. However, they would have some common required fields. The base schema would look something like:

```json
{
    "integration_name": {
        "type": "Job" | "Dataset",
        "namespace": [
            "uri_base",
            "auth",
            "host",
            "port"
        ],
        "name": [
            "db",
            "schema",
            "table"
        ],
        "example_namespace": "uri://auth.host:port",
        "example_name": "MY_DB.MY_SCHEMA.MY_TABLE",
        "example_unique_name": "uri://auth.host:port/MY_DB.MY_SCHEMA.MY_TABLE",
        "case_sensitive": true | false,
        "case": "upper" | "lower" | null
  }
}
```

In the above outline, the `integration_name` would match the name of the file. The `type` is one of "Job" or "Dataset". The `namespace` list contains all the elements, in order, needed to generate the namespace (this may be an issue if, like Redshift, there are multiple options. This might necessitate two files, one for each namespace). The `name` field is similar, but for elements to generate the name of the entity. Three examples are given, one for the namespace, another for the name, and a last for the unique name as a combination of the two, for clarity. Finally, a note about case sensitivity is included to ensure that all integrations are completely matching the spec.

In addition to the documents themselves, some testing framework should be developed (or tests simply added to integrations) to ensure that naming matches the structure and casing of the JSON in the files exactly.

**Comment by mobuchowski** [comment]
I think this is a good idea, but should be specified as JsonSchema. This would allow us to autogenerate relevant classes and methods for validation.

**Comment by pawel-big-lebowski** [comment]
That's a great idea 🚀  Could we use regular expressions with named groups instead? This would be a clear definition on how to extract elements from names and namespaces?

**Comment by denimalpaca** [comment]
@pawel-big-lebowski regex with named expressions instead of the name/namespace list? I agree that would be more ideal. I was also thinking about how that could be used to convert arbitrary URIs to unique names, then parse out a name / namespace from there.

@mobuchowski I'll look into JsonSchema, haven't done that before so didn't come to mind. That's how a lot of the rest of the OL project works, right? Like facet definitions?

**Comment by mobuchowski** [comment]
@denimalpaca yes, that's how rest of the OL specs work. We then generate Java class definitions from them - and want to generate Python ones too at some point.

**Comment by harels** [comment]
we're going to make a first pass on the clients side here #1873

**Comment by jenspfaug** [comment]
This is a good idea. Is this the right place to ask and maybe address why [Dataset namespaces](https://openlineage.io/blog/whats-in-a-namespace/#dataset-namespaces) contain a 'scheme' segment but [Job namespace](https://openlineage.io/blog/whats-in-a-namespace/#job-namespaces) do not? Scheme is useful for consumers to derive the source system (for a dataset) without further parsing any facets. This would also be useful for jobs.

**Comment by mountaincrab** [comment]
Is there any update on this? Regarding
> non-database systems, like Salesforce or Google Sheets
we are also investigating this and would like to produce lineage for these sources (and others e.g. Workday), but there doesn't seem to be any guidance on how to produce the namespace and name for datasets. I did ask a question related to this in the OpenLineage slack ([here](https://openlineage.slack.com/archives/C01CK9T7HKR/p1722408347459339)) the other day, before I saw this proposal!

**Comment by mobuchowski** [comment]
Update is that we should do that at some point. However, this would not help with defining names for custom names/namespaces but rather generate classes/functions/factory methods for standarized ones.

**Comment by mountaincrab** [comment]
Ah OK I see, so this is really just the "framework" for allowing proper documentation and code generation 👍

**Comment by mobuchowski** [comment]
Yes. Right now we do that for core schema objects - but as names & namespaces are oficially part of spec, we'd love to do this for them too.

**Comment by dolfinus** [comment]
Related: #3727
